File: dodam-application/dodam-rest-api/src/main/java/b1nd/dodam/restapi/club/application/ClubMemberUseCase.java
package b1nd.dodam.restapi.club.application;

import b1nd.dodam.domain.rds.club.entity.Club;
import b1nd.dodam.domain.rds.club.enumeration.ClubStatus;
import b1nd.dodam.domain.rds.club.enumeration.ClubTimeType;
import b1nd.dodam.domain.rds.club.service.ClubMemberService;
import b1nd.dodam.domain.rds.club.service.ClubService;
import b1nd.dodam.domain.rds.member.entity.Member;
import b1nd.dodam.domain.rds.member.entity.Student;
import b1nd.dodam.domain.rds.member.repository.StudentRepository;
import b1nd.dodam.restapi.auth.infrastructure.security.support.MemberAuthenticationHolder;
import b1nd.dodam.restapi.club.application.data.req.ClubPassReq;
import b1nd.dodam.restapi.club.application.data.req.JoinClubMemberReq;
import b1nd.dodam.restapi.club.application.data.res.*;
import b1nd.dodam.restapi.member.application.data.res.StudentWithImageRes;
import b1nd.dodam.restapi.support.data.Response;
import b1nd.dodam.restapi.support.data.ResponseData;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Component
@RequiredArgsConstructor
@Transactional(rollbackFor = Exception.class)
public class ClubMemberUseCase {
    private final ClubService clubService;
    private final ClubMemberService clubMemberService;
    private final MemberAuthenticationHolder authenticationHolder;
    private final StudentRepository studentRepository;

    public Response joinClub(JoinClubMemberReq req) {
        clubService.validateApplicationDuration(ClubTimeType.CLUB_APPLICANT);
        Student student = studentRepository.getByMember(authenticationHolder.current());
        clubMemberService.validateFirstGrade(student);
        clubMemberService.validateNoActiveCreativeClub(student);
        Club club = clubMemberService.findClubIfNotClubMember(req.clubId(), ClubStatus.ALLOWED, student, ClubStatus.DELETED);
        clubMemberService.saveAndValidateClubMembers(List.of(req.toEntity(student, club)));
        return Response.ok("동아리 입부 신청 성공");
    }

    public Response setClubMemberStatus(ClubPassReq req) {
        clubMemberService.validateByClubLeader(clubService.findById(req.clubId()), authenticationHolder.current());
        clubMemberService.setStatusStudentClub(req.studentId(), req.clubId(), req.status());
        return Response.ok("동아리 가입 상태 변경 성공");
    }

    public Response updateClubJoinRequestReceived(Long id, ClubStatus clubStatus) {
        clubMemberService.setClubMemberStatus(id,authenticationHolder.current(), clubStatus);
        return Response.ok("동아리 가입 업데이트 성공");
    }

    @Transactional(readOnly = true)
    public ResponseData<List<StudentWithImageRes>> getGradeStudents(boolean isSelf) {
        Member member = authenticationHolder.current();
        return ResponseData.ok("학년 불러오기 성공", isSelf
            ? clubMemberService.getAllGradeStudent(member).stream().map(StudentWithImageRes::of).toList()
            : clubMemberService.getSecondGradeStudent(member).stream().map(StudentWithImageRes::of).toList()
        );
    }

    @Transactional(readOnly = true)
    public ResponseData<List<ClubMemberRes>> getClubJoinRequestsReceived() {
        return ResponseData.ok("받은 부원 제안 불러오기 성공", clubMemberService.getJoinRequests(authenticationHolder.current(), ClubStatus.WAITING).stream().map(ClubMemberRes::of).toList());
    }

    @Transactional(readOnly = true)
    public ResponseData<ClubStudentRes> getClubLeader(Long id) {
        return ResponseData.ok("부장 불러오기 성공", ClubStudentRes.of(clubMemberService.getClubLeader(id)));
    }

    @Transactional(readOnly = true)
    public ResponseData<List<ClubJoinStudentRes>> getPendingClubMembers(Long id) {
        return ResponseData.ok("입부 희망 인원 불러오기 성공", clubMemberService.getStatusClubMembers(id, ClubStatus.PENDING).stream().map(ClubJoinStudentRes::of).toList());
    }

    @Transactional(readOnly = true)
    public ResponseData<ClubStudentListRes> getAllClubMembers(Long id) {
        return ResponseData.ok("동아리 멤버 불러오기 성공", getClubMembersByRole(id, clubMemberService.isClubLeader(id, authenticationHolder.current())));
    }

    public ResponseData<List<ClubMemberRes>> getMemberJoinRequests(int studentId) {
        Student student = studentRepository.getById(studentId);
        return ResponseData.ok("동아리 지망 불러오기 성공", clubMemberService.findAllCreativeClubByStudent(student).stream().map(ClubMemberRes::of).toList());
    }

    @Transactional(readOnly = true)
    public ResponseData<List<ClubMemberRes>> getStudentJoinRequest() {
        return ResponseData.ok("나의 동아리 입부 신청 불러오기 성공", clubMemberService.getJoinRequests(authenticationHolder.current(), ClubStatus.PENDING).stream().map(ClubMemberRes::of).toList());
    }

    @Transactional(readOnly = true)
    public ResponseData<List<ClubStatusRes>> getJoinedClubs() {
        return ResponseData.ok("나의 동아리 상태 불러오기 성공", clubMemberService.findUserAllowedClub(authenticationHolder.current()).stream().map(ClubStatusRes::of).toList());
    }

    @Transactional(readOnly = true)
    public ResponseData<List<ClubDetailRes>> getStudentClubStatus() {
        return ResponseData.ok("동아리 불러오기 성공", clubMemberService.getStudentClubStatus(studentRepository.getByMember(authenticationHolder.current())).stream().map(ClubDetailRes::of).toList());
    }

    private ClubStudentListRes getClubMembersByRole(Long id, boolean isLeader) {
        return ClubStudentListRes.of(
                isLeader,
                isLeader
                    ? clubMemberService.getAllClubMembers(id).stream().map(ClubStudentRes::of).toList()
                    : clubMemberService.getStatusClubMembers(id, ClubStatus.ALLOWED).stream().map(ClubStudentRes::of).toList()
        );
    }

}

File: services/service-club/src/main/kotlin/com/b1nd/dodamdodam/club/application/club/ClubMemberUseCase.kt
package com.b1nd.dodamdodam.club.application.club

import com.b1nd.dodamdodam.club.application.club.data.request.ClubPassRequest
import com.b1nd.dodamdodam.club.application.club.data.request.JoinClubMemberRequest
import com.b1nd.dodamdodam.club.application.club.data.response.ClubJoinStudentResponse
import com.b1nd.dodamdodam.club.application.club.data.response.ClubMemberResponse
import com.b1nd.dodamdodam.club.application.club.data.response.ClubStatusResponse
import com.b1nd.dodamdodam.club.application.club.data.response.ClubStudentListResponse
import com.b1nd.dodamdodam.club.application.club.data.response.ClubStudentResponse
import com.b1nd.dodamdodam.club.application.club.data.toClubMemberEntity
import com.b1nd.dodamdodam.club.domain.club.enumeration.ClubStatus
import com.b1nd.dodamdodam.club.domain.club.enumeration.ClubTimeType
import com.b1nd.dodamdodam.club.domain.club.service.ClubMemberService
import com.b1nd.dodamdodam.club.domain.club.service.ClubService
import com.b1nd.dodamdodam.club.infrastructure.grpc.MemberGrpcClient
import kotlinx.coroutines.runBlocking
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
@Transactional(rollbackFor = [Exception::class])
class ClubMemberUseCase(
    private val clubService: ClubService,
    private val clubMemberService: ClubMemberService,
    private val memberGrpcClient: MemberGrpcClient,
) {
    fun joinClubs(requests: List<JoinClubMemberRequest>, studentId: Long) {
        clubService.validateApplicationDuration(ClubTimeType.CLUB_APPLICANT)

        val members = requests.map { request ->
            val club = clubMemberService.findClubIfNotClubMember(
                clubId = request.clubId,
                state = ClubStatus.ALLOWED,
                studentId = studentId,
                excludeStatus = ClubStatus.DELETED,
            )
            request.toClubMemberEntity(studentId, club)
        }

        clubMemberService.saveAndValidateClubMembers(members)
    }

    fun setClubMemberStatus(request: ClubPassRequest, studentId: Long) {
        val club = clubService.findById(request.clubId)
        clubMemberService.validateByClubLeader(club, studentId)
        clubMemberService.setClubMemberStatus(request.clubId, request.studentId, request.status)
    }

    fun updateClubJoinRequestReceived(id: Long, studentId: Long, status: ClubStatus) {
        clubMemberService.setStatusStudentClub(studentId, id, status)
    }

    @Transactional(readOnly = true)
    fun getClubLeader(clubId: Long): ClubStudentResponse {
        val leader = clubMemberService.getClubLeader(clubId)
        val studentInfos = runBlocking {
            memberGrpcClient.getStudentsByIds(listOf(leader.studentId))
        }
        val studentInfo = studentInfos.firstOrNull()
        return ClubStudentResponse.fromEntity(
            clubMember = leader,
            name = studentInfo?.name ?: "",
            grade = studentInfo?.grade ?: 0,
            room = studentInfo?.room ?: 0,
            number = studentInfo?.number ?: 0,
            profileImage = studentInfo?.profileImage,
        )
    }

    @Transactional(readOnly = true)
    fun getPendingClubMembers(clubId: Long): List<ClubJoinStudentResponse> {
        val members = clubMemberService.getStatusClubMembers(clubId, ClubStatus.PENDING)
        val studentIds = members.map { it.studentId }
        val studentInfoMap = runBlocking {
            memberGrpcClient.getStudentsByIds(studentIds)
        }.associateBy { it.id }

        return members.map { member ->
            val studentInfo = studentInfoMap[member.studentId]
            ClubJoinStudentResponse.fromEntity(
                clubMember = member,
                name = studentInfo?.name ?: "",
                grade = studentInfo?.grade ?: 0,
                room = studentInfo?.room ?: 0,
                number = studentInfo?.number ?: 0,
                profileImage = studentInfo?.profileImage,
            )
        }
    }

    @Transactional(readOnly = true)
    fun getAllClubMembers(clubId: Long, studentId: Long): ClubStudentListResponse {
        val isLeader = clubMemberService.isClubLeader(clubId, studentId)
        val members = clubMemberService.getAllClubMembers(clubId)
        val studentIds = members.map { it.studentId }
        val studentInfoMap = runBlocking {
            memberGrpcClient.getStudentsByIds(studentIds)
        }.associateBy { it.id }

        val studentResponses = members.map { member ->
            val studentInfo = studentInfoMap[member.studentId]
            ClubStudentResponse.fromEntity(
                clubMember = member,
                name = studentInfo?.name ?: "",
                grade = studentInfo?.grade ?: 0,
                room = studentInfo?.room ?: 0,
                number = studentInfo?.number ?: 0,
                profileImage = studentInfo?.profileImage,
            )
        }

        return ClubStudentListResponse.of(isLeader, studentResponses)
    }

    @Transactional(readOnly = true)
    fun getMemberJoinRequests(studentId: Long): List<ClubMemberResponse> {
        val members = clubMemberService.findAllCreativeClubByStudent(studentId)
        return members.map { member ->
            val teacherName = member.club.teacherId?.let { teacherId ->
                runBlocking { memberGrpcClient.getTeacherById(teacherId) }?.name
            }
            ClubMemberResponse.fromEntity(member, teacherName)
        }
    }

    @Transactional(readOnly = true)
    fun getStudentJoinRequest(studentId: Long): List<ClubMemberResponse> {
        val members = clubMemberService.getJoinRequests(studentId, ClubStatus.PENDING)
        return members.map { member ->
            val teacherName = member.club.teacherId?.let { teacherId ->
                runBlocking { memberGrpcClient.getTeacherById(teacherId) }?.name
            }
            ClubMemberResponse.fromEntity(member, teacherName)
        }
    }

    @Transactional(readOnly = true)
    fun getJoinedClubs(studentId: Long): List<ClubMemberResponse> {
        val members = clubMemberService.getJoinRequests(studentId, ClubStatus.ALLOWED)
        return members.map { member ->
            val teacherName = member.club.teacherId?.let { teacherId ->
                runBlocking { memberGrpcClient.getTeacherById(teacherId) }?.name
            }
            ClubMemberResponse.fromEntity(member, teacherName)
        }
    }

    @Transactional(readOnly = true)
    fun getStudentClubStatus(studentId: Long): List<ClubStatusResponse> {
        val members = clubMemberService.getJoinRequests(studentId, ClubStatus.PENDING) +
            clubMemberService.getJoinRequests(studentId, ClubStatus.ALLOWED)
        return members.map { member ->
            ClubStatusResponse.fromEntity(member)
        }
    }
}
